firestore/bulkwriter.go

```diff
@@ -0,0 +1,319 @@
+}
+
+// newBulkWriter creates a new instance of the BulkWriter. This
+// version of BulkWriter is intended to be used within go routines by the
+// callers.
+func newBulkWriter(ctx context.Context, c *Client, database string) *BulkWriter {
+	// Although typically we shouldn't store Context objects, in this case we
+	// need to pass this Context through to the Bundler
+	ctx, cancel := context.WithCancel(withResourceHeader(ctx, c.path()))
+
+	bw := &BulkWriter{
+		database:        database,
+		start:           time.Now(),
+		vc:              c.c,
+		isOpen:          true,
+		maxOpsPerSecond: defaultStartingMaximumOpsPerSecond,
+		docUpdatePaths:  make(map[string]bool),
+		cancel:          cancel,
+		ctx:             ctx,
+	}
+
+	// can't initialize with struct above; need instance reference to BulkWriter.send()
+	bw.bundler = bundler.NewBundler(BulkWriterJob{}, bw.send)
+	bw.bundler.HandlerLimit = bw.maxOpsPerSecond
+	bw.bundler.BundleCountThreshold = maxBatchSize
+	bw.bundler.BundleByteLimit = 0 // unlimited size
+
+	return bw
+}
+
+// End sends all enqueued writes in parallel and closes the BulkWriter to new requests.
+// After calling End(), calling any additional method automatically returns
+// with an error. This method completes when there are no more pending writes
+// in the queue.
+func (b *BulkWriter) End() {
+	b.Flush()
+	b.openLock.Lock()
+	b.isOpen = false
+	b.openLock.Unlock()
+}
+
+// Flush commits all writes that have been enqueued up to this point in parallel.
+// This method blocks execution.
+func (bw *BulkWriter) Flush() {
+	bw.bundler.Flush()
+}
+
+// IsOpen gets the current open or closed state of the BulkWriter.
+func (bw *BulkWriter) IsOpen() bool {
+	bw.openLock.Lock()
+	defer bw.openLock.Unlock()
+	return bw.isOpen
+}
+
+// Create adds a document creation write to the queue of writes to send.
+// Note: You cannot write to (Create, Update, Set, or Delete) the same document more than once.
+func (bw *BulkWriter) Create(doc *DocumentRef, datum interface{}) (*BulkWriterJob, error) {
+	err := bw.checkWriteConditions(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(telpirion): Talk to FS team about why we're doing this
+	w, err := doc.newCreateWrites(datum)
+	if err != nil {
+		return nil, fmt.Errorf("firestore: cannot create %v with %v", doc.ID, datum)
+	}
+
+	if len(w) > 1 {
+		return nil, fmt.Errorf("firestore: too many document writes sent to bulkwriter")
+	}
+
+	j := bw.write(w[0])
+	return j, nil
+}
+
+// Delete adds a document deletion write to the queue of writes to send.
+// Note: You cannot write to (Create, Update, Set, or Delete) the same document more than once.
+func (bw *BulkWriter) Delete(doc *DocumentRef, preconds ...Precondition) (*BulkWriterJob, error) {
+	err := bw.checkWriteConditions(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := doc.newDeleteWrites(preconds)
+	if err != nil {
+		return nil, fmt.Errorf("firestore: cannot delete doc %v", doc.ID)
```

Comment:
Should the underlying error be wrapped for context? Applies for other methods too.

Reply:
Done.